tck/src/main/java/org/eclipse/microprofile/lra/tck/service/spi/LRARecoveryService.java

```diff
@@ -0,0 +1,88 @@
+     * are verified so the implementation must make sure that this method doesn't return before the validation
+     * can happen.
+     * </p>
+     * 
+     * Example:
+     * <ol>
+     *   <li>call to close LRA with id lraId (Complete calls are starting to be performed)</li>
+     *   <li>call to {@link LRARecoveryService#waitForCallbacks(URI)} with lraId -- wait for all Complete calls to be
+     * delivered and optionally perform some implementation specific actions to make the calls return faster</li>
+     *   <li>call to {@link LRARecoveryService#waitForCallbacks(URI)} returns</li>
+     *   <li>the TCK test verifies that all Complete methods have been called</li>
+     * </ol>
+     *
+     * @param lraId the LRA id of the LRA which callbacks should be delivered when this method exits
+     * @throws Exception in case of any error. The exception will be propagated to fail the enclosing TCK test.
+     */
+    void waitForCallbacks(URI lraId) throws Exception;
+
+    /**
+     * <p>
+     * This method is utilized by the TCK to wait for the recovery of a particular LRA
+     * (replay of Status, Complete, Compensate) once. This means that this method should be called when the passed
+     * LRA is in one of its intermediate states (Completing, Compensating). The invocation should retry the calls
+     * to all enlisted participant methods according to the specification and wait for their completion (meaning the
+     * delivery of the participants responses back to the implementation). The implementation is allowed to perform
+     * any implementation specific actions to make the recovery perform faster (e.g. immediate recovery trigger).
+     * When this method returns the participant methods invocations results are verified so the implementation must
+     * make sure that the method doesn't return before the validation can happen.
+     * </p>
+     * 
+     * Example:
+     * <ol>
+     *   <li>call to close LRA with id lraId (Complete calls are starting to be performed)</li>
+     *   <li>one or more of the enlisted participants returns HTTP status code 202 from its
+     *   completion notification which implies the participant is Completing</li>
+     *   <li>call to {@link LRARecoveryService#waitForRecovery(URI)} with the lraId -- wait for the replay
+     * of the Complete (or Status) call and for its delivery, and optionaly perform some implementation
+     * specific actions to make the recovery execution faster</li>
+     *   <li>call to {@link LRARecoveryService#waitForRecovery(URI)} returns</li>
+     *   <li>the TCK test verifies that Complete (or Status) have been called (retried)</li>
+     * </ol>
+     * 
+     * @param lraId the LRA id of the LRA which should be recovered when this method exits
+     * @throws Exception in case of any error. The exception will be propagated to fail the enclosing TCK test.
+     *              
+     */
+    void waitForRecovery(URI lraId) throws Exception;
+}
```

Comment:
I am finding it difficult to parse the javadoc. Your first attempt was better, I took that a varied it a little- how about the following alternative:

```java
/**
 *     The LRA model guarantees eventual consistency but does not say when
 *     participants will be brought into a consistent state. When a request
 *     to close/cancel an LRA is received, the implementation is allowed to
 *     delay sending out the complete/compensate callbacks (the specification
 *     only requires that an implementation must eventually send them).
 *
 *     This SPI provides a mechanism for implementations to indicate when
 *     it knows that callbacks have been sent and responses received or when
 *     callbacks have been issued and all participants have reached an end state.
 *
 */
public interface LRARecoveryService {
    /**
     * Wait for the delivery of all participant callbacks (Complete or Compensate).
     * When this method returns the caller can be certain that the callbacks were
     * sent and responses received (including error responses).
     *
     * @param lraId the LRA context
     * @throws LRACallbackException the implementation was unable to determine whether
     * or not the callbacks were received by all participants
     */
    void waitForCallbacks(URI lraId) throws CallbackException;

    /**
     * Waits for all participants to reach an end state.
     * Question do we also want the method to wait for any AfterLRA notifications to be successfully delivered?
     *
     * @param lraId the LRA context
     * @throws LRACallbackException the implementation was unable to determine whether
     * or not all participants have reached an end state.
     */
    void waitForRecovery(URI lraId) throws CallbackException;
}
```

Reply:
so I see, we had a misunderstanding - with waitForRecovery you wanted to change the definition  of what triggerRecovery does, not just a rename. Ok, I will use this javadoc and fix your other comment.

Reply:
>     /**
>      * ...
>      * Question do we also want the method to wait for any AfterLRA notifications to be successfully delivered?

@xstefank Did you also see my question in the javadoc. Is that something we need and if so can it be implicit in the `waitForRecovery` method?

Reply:
yes, I included also a requirement for afterLRA to be delivered.

Reply:
@mmusgrov but if we are going with your definition (Waits for all participants to reach an end state) we are unable to test intermediate states of LRAs -- check the testForget() TCK test, please.

Reply:
So go with some similar to [one of my earlier suggestions](https://github.com/eclipse/microprofile-lra/issues/249#issuecomment-549723582).

Reply:
> yes, I included also a requirement for afterLRA to be delivered.

We need good javadoc so that it is clear what the implementer's responsibilities are.

Reply:
> @mmusgrov but if we are going with your definition (Waits for all participants to reach an end state) we are unable to test intermediate states of LRAs -- check the testForget() TCK test, please. So, are we going back to triggerRecovery() or should I change the tests?

I did not fully understand your javadoc so was unable to determine what the responsibilities were. That is why I asked if I could propose the alternative text.

Can you explain what we need in order to verify the `testForget()` TCK test.

Reply:
it would be better to check the code directly. testForget closes LRA which goes into Closing state, then we "triggerRecovery" so the Status method is invoked. If we now block until the participant reaches the final state, this test will never end. The state is cleared after som asserts and then the recovery is triggered again. So we can either have `triggerRecovery(URI)` and `waitForRecovery(URI)` or we can replace all `waitForRecovery(URI)` calls with `triggerRecovery(URI); waitForCallbacks(URI);`. Or of course we can change this test, but then we will be always be able to test only end states which I guess is not something we want.

Reply:
So don't block